jax/_src/ops/scatter.py

```diff
@@ -15,11 +15,21 @@
 # Helpers for indexed updates.
 
 
+from typing import Optional, Sequence, Tuple, Union
+
+import numpy as np
+
 from jax import lax
 from jax._src.numpy import lax_numpy as jnp
 from jax._src import util
 
 
+Index = Union[int, slice, Sequence[int], np.ndarray, 'ellipsis',
+              Tuple[Union[int, slice, Sequence[int], np.ndarray, 'ellipsis'], ...]]
```

Comment:
This also needs `None` to account for `np.newaxis`.

It might be more readable to express it as something like:
```python
SingleIndex = Union[int, slice, Sequence[int], np.ndarray, 'ellipsis', None]
Index = Union[SingleIndex, Tuple[SingleIndex, ...]]
```

Reply:
I like the composite definition.  Very nice.

I'm not sure about `None` though.  Which scatter function supports adding axes?  It appears that not only do they not add axes, but they sliently ignore them!

```python
z = jnp.zeros((4, 4))
index_add(z, (None, ...), 1).shape  # (4, 4)!!
```
Looks like a bug to be honest.

Reply:
I think by analogy to numpy's index update, it makes sense that `None` is a valid input, and should be ignored:
```python
import numpy as np
x = np.arange(12).reshape(3, 4)
x[:, 1, None] += 10
print(x)
# array([[ 0, 11,  2,  3],
#        [ 4, 15,  6,  7],
#        [ 8, 19, 10, 11]])

import jax.numpy as jnp
x = jnp.arange(12).reshape(3, 4)
print(x.at[:, 1, None].add(10))
# [[ 0 11  2  3]
#  [ 4 15  6  7]
#  [ 8 19 10 11]]
```

Reply:
Okay, great!  I added `None` to `SingleIndex` to match the behavior.

Reply:
It turns out the `None` is not ignored; for example, this fails without it:
```python
jnp.arange(4).at[:, None].max(jnp.ones((4, 1)))
```

Reply:
Yes, good point!

Reply:
I think the key is that the update value is broadcast to the index shape. Adding a leading `None` to the index doesn't  do a whole lot because broadcasting implicitly does that.